internal/proxy/task_query.go

```diff
@@ -409,46 +409,66 @@ func IDs2Expr(fieldName string, ids *schemapb.IDs) string {
 	return fieldName + " in [ " + idsStr + " ]"
 }
 
-func mergeRetrieveResults(ctx context.Context, retrieveResults []*internalpb.RetrieveResults) (*milvuspb.QueryResults, error) {
-	var ret *milvuspb.QueryResults
-	var skipDupCnt int64
-	var idSet = make(map[interface{}]struct{})
+func reduceRetrieveResults(ctx context.Context, retrieveResults []*internalpb.RetrieveResults, queryParams *queryParams) (*milvuspb.QueryResults, error) {
+	log.Ctx(ctx).Debug("reduceInternelRetrieveResults", zap.Int("len(retrieveResults)", len(retrieveResults)))
+	var (
+		ret = &milvuspb.QueryResults{}
+
+		skipDupCnt int64
+		loopEnd    int
+	)
 
-	// merge results and remove duplicates
-	for _, rr := range retrieveResults {
-		numPks := typeutil.GetSizeOfIDs(rr.GetIds())
-		// skip empty result, it will break merge result
-		if rr == nil || rr.Ids == nil || rr.GetIds() == nil || numPks == 0 {
+	validRetrieveResults := []*internalpb.RetrieveResults{}
+	for _, r := range retrieveResults {
+		size := typeutil.GetSizeOfIDs(r.GetIds())
+		if r == nil || len(r.GetFieldsData()) == 0 || size == 0 {
 			continue
 		}
+		validRetrieveResults = append(validRetrieveResults, r)
+		loopEnd += size
+	}
+
+	if len(validRetrieveResults) == 0 {
+		return ret, nil
+	}
 
-		if ret == nil {
-			ret = &milvuspb.QueryResults{
-				FieldsData: make([]*schemapb.FieldData, len(rr.FieldsData)),
+	ret.FieldsData = make([]*schemapb.FieldData, len(validRetrieveResults[0].GetFieldsData()))
+	idSet := make(map[interface{}]struct{})
+	cursors := make([]int64, len(validRetrieveResults))
+
+	if queryParams != nil && queryParams.limit != typeutil.Unlimited {
+		loopEnd = int(queryParams.limit)
+
+		if queryParams.offset > 0 {
```

Comment:
if user don't set limit, then offset won't be handled?
doesn't make sense to me